### PR TITLE
Add new CI Spellcheck job

### PR DESCRIPTION
**Pull request by Owen-CH-Leung**
As per #6263 , This PR introduce a new CI spellcheck job using `cargo spellcheck`. The job leverages the `spellcheck.toml` and `spellcheck.dic` to perform spelling check on the docs.

### CODE REVIEW

**Comment by maminrayej** [comment]
I think it makes sense to surround `Rust` and `tokio` specific words with backticks. This would significantly reduce the size of the `.dic` file, and makes it more maintainable. Also, we don't have to fight the spellchecker each time a new struct, method, etc is introduced.
Some examples of Rust specific words: `&str`, `&mut`, ...
Some examples of tokio specific words: `AsHandle`, `AsyncWrite`, ...

Additionally, it would be beneficial to incorporate a section in `CONTRIBUTING.md` that outlines the spell checking process. Similar to existing sections on `rustfmt`, `clippy`, and testing, this new section could provide guidelines on running the spell check command, updating the dictionary, and documenting in a way that aligns with spellchecker requirements, such as the use of backticks.

**Comment by Owen-CH-Leung** [comment]
> maminrayej

Thanks for your feedback. So you mean we should modify `Rust` and `tokio` specific words inplace so that we don't have to add the words in `spellcheck.dic` ? 

e.g. `&mut` in https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/time/entry.rs#L85-L88 should be changed inplace to ``&mut``

**Comment by maminrayej** [comment]
Yes. In that particular example I think `&mut TimerEntry` is more reasonable. Basically anywhere referring to code I  believe it should be enclosed in backticks because it will most probably be flagged by `hunspell`.

**Comment by Owen-CH-Leung** [comment]
Ok make sense to me. It'll require some more time to modify each comment piece by piece though. I'll revise them soon

**Comment by maminrayej** [comment]
No worries. Thanks.

**Comment by Owen-CH-Leung** [comment]
@maminrayej Can I ask for your review again ? I backtick-ed lots of rust / tokio words and trimmed the dictionary file down to 3xx entries. 

However, I notice that not all rust / tokio words can be backtick-ed. (e.g. `AsRawHandle` cannot be backtick-ed and removed since it's part of the doc test in [here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/net/windows/named_pipe.rs#L2096). And I don't think we can configure `cargo spellcheck` to skip checking the doctest

**Comment by maminrayej** [comment]
Great progress!

The issue you've mentioned happens because of a formatting mistake in that file. There are two empty lines that are commented using `//` instead of `///`. One at line [2062](https://github.com/tokio-rs/tokio/blob/4846959e8a534cf2aff63f8613e5f69a7d2ce5f8/tokio/src/net/windows/named_pipe.rs#L2062), and the other one at line [2097](https://github.com/tokio-rs/tokio/blob/4846959e8a534cf2aff63f8613e5f69a7d2ce5f8/tokio/src/net/windows/named_pipe.rs#L2097). I believe this confuses `cargo spellcheck`. If you fix these two lines, that file can be successfully spellchecked with this small `.dic` file:
```
1
dwOpenMode
Tokio
tokio
runtime
Runtime
syscall
Dev
wakeup
waker
async
qos
sync
unsafety
```
I've tested this using the following command:
```
cargo spellcheck check tokio/src/net/windows/named_pipe.rs
```
Keep an eye out for these formatting mistakes if you encounter something like this again. When you apply these fixes, the size of the `.dic` file could be further reduced.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -194,6 +194,16 @@ MIRIFLAGS="-Zmiri-disable-isolation -Zmiri-tag-raw-pointers" \
...
 
+You can also perform spell-check on tokio codebase. For details of how to use the spellcheck tool, feel free to visit
+https://github.com/drahnr/cargo-spellcheck
+```
+# First install the spell-check plugin
+cargo install --locked cargo-spellcheck
+
+# Then run the cargo spell check command
+cargo spell check
+```
```
Comment (maminrayej) [inline CONTRIBUTING.md:205]: I'd add a section for it. Also, please explain what to do if `cargo-spellcheck` rejects a word: 
If it's code related, it should be enclosed with backticks. If not, they should add it to the `.dic` file.

Comment (maminrayej) [reply CONTRIBUTING.md:205]: Also the command should be:
```
cargo spellcheck check
```

File: .github/workflows/ci.yml
Comment (Darksonn) [inline .github/workflows/ci.yml:1]: These files are currently triggering github's warning for a missing newline at the end. Please add those newlines.

File: spellcheck.dic [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+>
+adaptor
+adaptors
+Adaptors
+AIO
+ambiant
+amongst
+api
+APIs
+assert_eq
```
Comment (maminrayej) [inline spellcheck.dic:37]: code related

File: spellcheck.dic [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+backend
+backpressure
+backtrace
+backtraces
+backtracing
+binded
+bitfield
+bitfields
+Blockingly
+bool
```
Comment (maminrayej) [inline spellcheck.dic:49]: code related

File: spellcheck.dic [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+boolean
+broadcasted
+buf
```
Comment (maminrayej) [inline spellcheck.dic:52]: likely code related

File: spellcheck.dic [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+cancelled
+cancelling
+Cancelling
+cfg
```
Comment (maminrayej) [inline spellcheck.dic:56]: likely code related

File: spellcheck.dic [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+cloneable
+codebase
+codec
+codecs
+combinator
+combinators
+Config
+config
+connectionless
+const
```
Comment (maminrayej) [inline spellcheck.dic:67]: likely code related

File: spellcheck.dic [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+filename
+filesystem
+filesystems
+fn
+fns
+FreeBSD
+frontend
+fs
+functionalities
+fuzz_linked_list
```
Comment (maminrayej) [inline spellcheck.dic:116]: code related

File: spellcheck.dic [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+multicast
+Multithreaded
+mut
+mutex
+Mutex
+Nagle
+nonblocking
+nondecreasing
+noop
+notitifications
```
Comment (maminrayej) [inline spellcheck.dic:170]: I think this is a misspelling in the documentation. It should be notifications.

File: spellcheck.dic [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+runtime
+runtimes
+runtime's
+rwlock
+rx
+scalability
+scheduler's
+semver
+setpgid
+shard_id
```
Comment (maminrayej) [inline spellcheck.dic:210]: code related

**Review by maminrayej — CHANGES_REQUESTED** [review]

**Comment by Owen-CH-Leung** [comment]
Thanks @maminrayej for your comment. I've backticked some more words and removed them from the dictionary. However, there're still a few words which I'd say it's just normal english words and I'd recommend putting it into dictioary (e.g. `amongst` ([here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/scheduler/multi_thread/worker.rs#L683)), `Blockingly` ([here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/sync/rwlock.rs#L473)), `broadcasted` ([here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/sync/broadcast.rs#L408)), `cancelled`([here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/scheduler/multi_thread/worker.rs#L48)), `functionalities`([here](https://github.com/tokio-rs/tokio/blob/master/tokio/src/sync/broadcast.rs#L498)) etc...)

**Comment by maminrayej** [comment]
Ah I think it's a misunderstanding. GitHub shows around the place where a comment is placed. For example in here:
```
scheduler's
semver
setpgid
shard_id
```
Comment is placed on `shard_id`, but a few lines above it is also displayed. I just meant that `shard_id` is code related.

Let me give you the list of words here:
```
assert_eq
bool
buf
cfg
const
fuzz_linked_list
shard_id
```

Everything else should be reverted back. Sorry for the inconvenience.

**Comment by Owen-CH-Leung** [comment]
> assert_eq
> bool
> buf
> cfg
> const
> fuzz_linked_list
> shard_id

Thanks. Reverted everything and removed only these 7 words

**Comment by Owen-CH-Leung** [comment]
Can you also trigger the failing test to rerun ? I saw the `cross-test-without-parking_lot` was failing but I don't think my changes break the codebase...I suspect it's just intermittent failure

**Review by maminrayej — APPROVED** [review]

**Comment by maminrayej** [comment]
Thanks!

@Darksonn I believe this PR is in good shape. There are additional work that can be done like making sure the `.dic` file is ordered and it doesn't contain duplicates. But I think it's better to do it in a different PR.

**Pull request merged**